include/sys/sys_heap.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_SYS_SYS_HEAP_H_
+#define ZEPHYR_INCLUDE_SYS_SYS_HEAP_H_
+
+#include <kernel.h>
+
+/* Simple, fast heap implementation.
+ *
+ * A more or less conventional segregated fit allocator with
+ * power-of-two buckets.
+ *
+ * Excellent space efficiency.  Chunks can be split arbitrarily in 8
+ * byte units.  Overhead is only four bytes per allocated chunk (eight
+ * bytes for heaps >256kb or on 64 bit systems), plus a log2-sized
+ * array of 2-word bucket headers.  No coarse alignment restrictions
+ * on blocks, they can be split and merged (in units of 8 bytes)
+ * arbitrarily.
+ *
+ * Simple API.  Initialize at runtime with any blob of memory and not
+ * a macro-generated, carefully aligned static array.  Allocate and
+ * free by user pointer and not an opaque block handle.
+ *
+ * Good fragmentation resistance.  Freed blocks are always immediately
+ * merged with adjacent free blocks.  Allocations are attempted from a
+ * sample of the smallest bucket that might fit, falling back rapidly
+ * to the smallest block guaranteed to fit.  Split memory remaining in
+ * the chunk is always returned immediately to the heap for other
+ * allocation.
+ *
+ * Excellent performance with firmly bounded runtime.  All operations
+ * are constant time (though there is a search of the smallest bucket
+ * that has a compile-time-configurable upper bound, setting this to
+ * extreme values results in an effectively linear search of the
+ * list), objectively fast (~hundred instructions) and and amenable to
+ * locked operation.
+ */
+
+/* Note: the init_mem/bytes fields are for the static initializer to
+ * have somewhere to put the arguments.  The actual heap metadata at
```

Comment:
Hm... forgot to remove this bit.  Ignore this comment and those fields.  I yanked the static initializer from this version entirely because the semantics of SYS_MEM_POOL_DEFINE is a little too weird to wrap anyway.